heron/spi/src/java/com/twitter/heron/spi/common/Config.java

```diff
@@ -44,17 +47,31 @@ public static Builder newBuilder(boolean loadDefaults) {
   }
 
   public static Config expand(Config config) {
+    return expand(config, 0);
+  }
+
+  private static Config expand(Config config, int previousTokensCount) {
     Config.Builder cb = Config.newBuilder();
+    int tokensCount = 0;
     for (String key : config.getKeySet()) {
       Object value = config.get(key);
       if (value instanceof String) {
         String expandedValue = Misc.substitute(config, (String) value);
+        if (expandedValue.contains("${")) {
+          tokensCount++;
+        }
         cb.put(key, expandedValue);
       } else {
         cb.put(key, value);
       }
     }
-    return cb.build();
+    if (previousTokensCount != tokensCount) {
```

Comment:
Each time the method is invoked we iterate through each config value and do round of pattern substitution on the value, but substituting patterns can in the value can result in more tokens. For example `${HERON_BIN}/heron-executor` can be changed to `${HERON_HOME}/bin/heron-executor`, so after token substitution we can still have values with tokens. `tokensCount` counts the fields that still have tokens to know if we need to recurse for another pass.

I'm curious though about how this worked before without recursion so I'm going to investigate some more. If we do in fact need it, I'll provide more docs about what's happening.